docs: Add a developers guide to testsys (#499)
bottlerocket-os/bottlerocket-test-system: DEVELOPER.md

# Developer Guide for Testsys

## What is testsys

> Testsys is a framework that leverages kubernetes (k8s) to create resources and run tests.

Reviewer: nit: capitalize Kubernetes
```diff
- Testsys is a framework that leverages kubernetes (k8s) to create resources and run tests.
+ Testsys is a framework that leverages Kubernetes (k8s) to create resources and run tests.
```

#### [Dockerfile](Dockerfile)

Testsys relies on docker images to run tests and create resources.
The Dockerfile contains the logic needed to create all bottlerocket related images.

#### [Agent](agent)

Testsys is broken up into 3 main things: the controller, test agent, and resource agent.

[agent-common](agent/agent-common) contains logic that is related to both the test and resource agents.

[test-agent](agent/test-agent) provides the framework for creating test agents.
To create a test agent, the [`Runner` trait](agent/test-agent/src/lib.rs#L51) needs to be implemented.
Check out the [example test agent](agent/test-agent/examples/example_test_agent/main.rs) to see how to create a test agent.

[resource-agent](agent/resource-agent) provides the framework for creating resource agents.
To create a resource agent, the [`Create` trait](agent/resource-agent/src/mod.rs#L37) and [`Destroy` trait](agent/resource-agent/src/mod.rs#L73) need to be implemented.
Check out the [example resource agent](agent/resource-agent/examples/example_resource_agent/main.rs) to see how to create a resource agent.

> 

Reviewer: The above section is confusing for me:

> test-agent provides the framework for creating test agents.

and

> resource-agent provides the framework for creating resource agents.

tell me that these two things are frameworks but they are self describing definitions. I.e. I'd expect these to describe _what_ a `test-agent` and/or `resource-agent` are without using the term itself in the definition.

### [Bottlerocket](bottlerocket)

The bottlerocket directory of the repo is broken into 3 sections: agents, testsys, and types.

[agents](bottlerocket/agents) contains the implementations of all bottlerocket agents.

The types used by the agents can be found in [types](bottlerocket/types).

[testsys](bottlerocket/testsys) contains a cli implementation that is closely related to bottlerocket agents but will be removed in the near future.

### [Cli](cli)

> Cli is and example command line interface that uses the [TestManager](model/src/test_manager/manager.rs) to interact with a testsys cluster.

Reviewer: Typo:

```diff
- Cli is and example command line interface that uses the [TestManager](model/src/test_manager/manager.rs) to interact with a testsys cluster.
+ Cli is an example command line interface that uses the [TestManager](model/src/test_manager/manager.rs) to interact with a testsys cluster.
```

### [Controller](controller)

> The controller orchestrates the the interaction between test and resource. For more info see the [design](design/DESIGN.md) document. 

Reviewer: Double word:

```diff
- The controller orchestrates the the interaction between test and resource. For more info see the [design](design/DESIGN.md) document.
+ The controller orchestrates the interaction between test and resource. For more info see the [design](design/DESIGN.md) document.
```

This is a bit confusing for me as well. What is a test and a resource in this context?

### [Model](model)

The model library contains the infrastructure that testsys is built upon.

> This includes the crds for [tests](model/src/test.rs) and [resources](model/src/resource.rs) as well as [everything](model/system) needed to be installed to the testsys cluster.

Reviewer: Nit: [CRDs should be capitilized](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) as an acronym

```diff
- This includes the crds for [tests](model/src/test.rs) and [resources](model/src/resource.rs) as well as [everything](model/system) needed to be installed to the testsys cluster.
+ This includes the CRDs for [tests](model/src/test.rs) and [resources](model/src/resource.rs) as well as [everything](model/system) needed to be installed to the testsys cluster.
```

To see how to use the TestManager look at the [example cli](cli).

Model also contains a [TestClient](model/src/clients/test_client) for simple interactions with `Test`s and a [ResourceClient](model/src/clients/resource_client) for `Resource`s.

### [Selftest](selftest)

Selftest is a library used to validate incoming pull request by testing the overall testsys framework.

## Getting Started

Testsys is currently tested on `x86_64` machines, these steps may not work as expected on other operating systems.

### Setting up testsys

#### Create a testsys directory
Create a directory that you wish use as a working directory and navigate to it. 
The example provided creates a directory in the users home, but a directory may be created anywhere.

```bash
cd ~
mkdir testsys && cd testsys
pwd 
```

#### Clone the testsys repo
While in the testsys directory, clone the `bottlerocket-test-system` repo from github.

```bash
git clone https://github.com/bottlerocket-os/bottlerocket-test-system.git
cd bottlerocket-test-system
export TESTSYS_DIR=$(pwd)
```

The testsys directory is now set up.

### Installing tools

#### Install Kind

> Instead of creating a long-standing eks cluster for testsys development, kind allows the quick creation and destruction of k8s clusters which helps improve the testing experience.

Reviewer: It may be useful to indicate that `kind` is for _local_ clusters that exist within the developers local Docker

Follow the [installation instructions](https://kind.sigs.k8s.io/docs/user/quick-start/) to set up kind.

#### Install Kubectl
While all testing user interactions can be done with `testsys`, it is helpful to look at test, resource, and pod objects in a k8s cluster with `kubectl`.
Follow the [installation instructions]( https://kubernetes.io/docs/tasks/tools/) to set up kubectl.

#### Install Docker
Docker is required for building testsys images if docker is not installed follow the [installation instructions](https://docs.docker.com/get-docker/).

## Running a test

### Navigate to the testsys directory
```bash
cd ${TESTSYS_DIR}
```

### Build testing images
First, the images need to be created and tagged so they can be loaded into a kind cluster.
```bash
make controller
make example-test-agent
docker tag controller controller:demo
docker tag example-test-agent example-test-agent:demo
```
This step uses the Makefile targets for building the controller image and the example test agent image. The images then need to be tagged so that the cluster uses loaded images instead of ones from external sources.

### Install the testsys cli
Install the `cli` from the repo.
```bash

> cargo install --path cli

Reviewer: This installs the testsys CLI as literally `cli` in my cargo bin:

```
❯ which cli
~/.cargo/bin/cli
❯ cli help
cli 0.0.1
The command line interface for setting up a Bottlerocket TestSys cluster and running tests
```

I'm not really seeing a way for `cargo install` to name the crate something different. Maybe we can look at naming this something more sensible in a different issue? Sounds like this CLI will also soon be moving to the core bottlerocket repo?

Author: The testsys cli is moving to the bottlerocket repo. Cli was created as an example cli for anyone wanting to use the testsys framework. `test-cli` may be a better name for it.

Reviewer: Good to know! Probably worth having discussing outside of this pull request. But thanks for the context!

```

### Create the test yaml file
The yaml file tells the controller and the test agent how the test should be run.
```bash
echo '---
apiVersion: testsys.bottlerocket.aws/v1
kind: Test
metadata:
  name: hello-bones
  namespace: testsys-bottlerocket-aws
spec:
  agent:
    name: hello-agent
    image: "example-test-agent:demo"
    keepRunning: false
    configuration:
      mode: Fast
      person: Bones the Cat
      helloCount: 3
      helloDurationMilliseconds: 500
  resources: []
  dependsOn: []' > example_test_agent.yaml
```

> 

Reviewer: Nit: extra line here?

*Note: If a kind cluster is being created and destroyed each time, there is no need to manually delete the tests.*

### Delete the kind cluster
```bash
kind delete cluster --name testsys-demo
```

## Running a resource

### Navigate to the testsys directory
```bash
cd ${TESTSYS_DIR}
```

### Build testing images
First, the images need to be created and tagged so they can be loaded into a kind cluster.
```bash
make controller
make example-test-agent
make duplicator-resource-agent
docker tag controller controller:demo
docker tag example-test-agent example-test-agent:demo
docker tag duplicator-agent duplicator-agent:demo
```
This step uses the Makefile targets for building the controller image, the duplicator resource agent image, and the example test agent image. The images then need to be tagged so that the cluster uses loaded images instead of ones from external sources.

### Install the testsys cli
Install the `cli` from the repo. 
*This step can be skipped if the cli is already installed.*
```bash
cargo install --path cli
```

### Create the resource and test yaml file

> The yaml file describes 2 crds that will be created.

Reviewer: Nit: capitalize CRDs

To delete all tests and resources from the cluster use `cli delete`.
Another option is to remove everything from the TestSys namespace using `cli uninstall`.
This will delete the TestSys namespace and everything inside it, including the controller, secrets, tests and resources.

*Note: If a `kind` cluster is being created and destroyed each time, there is no need to manually delete the tests.*

### Delete the `kind` cluster
```bash
kind delete cluster --name testsys-demo
```

## Running a resource

### Navigate to the TestSys directory
```bash
cd ${TESTSYS_DIR}
```

### Build testing images
First, the images need to be created and tagged so they can be loaded into a `kind` cluster.
```bash
make controller
make example-test-agent
make duplicator-resource-agent
docker tag controller controller:demo
docker tag example-test-agent example-test-agent:demo
docker tag duplicator-resource-agent duplicator-resource-agent:demo
```
This step uses the Makefile targets for building the controller image, the duplicator resource agent image, and the example test agent image. The images then need to be tagged so that the cluster uses loaded images instead of ones from external sources.

> ### Install the TestSys cli

Reviewer: ```suggestion
### Install the TestSys CLI
```

nit: CLI is an acronym here and should be capitalized

kind: Test
metadata:
  name: hello-bones
  namespace: testsys-bottlerocket-aws
spec:
  agent:
    name: hello-agent
    image: "example-test-agent:demo"
    keepRunning: false
    configuration:
      mode: Fast
      person: Bones the Cat
      helloCount: {duplicator.info}
      helloDurationMilliseconds: 500
  resources: [duplicator]
  dependsOn: []' > example.yaml
```

### Create the `kind` cluster
Check if the `kind` cluster already exists.
```bash
kind get clusters
```

If the cluster already exists, it should be deleted.
```bash
kind delete cluster --name testsys-demo
```

Create the new demo cluster.
```bash
kind create cluster --name testsys-demo
```

Now the images created earlier need to be added to the cluster.
```bash
kind load docker-image controller:demo example-test-agent:demo duplicator-resource-agent:demo --name testsys-demo
```

### Install TestSys
Before any tests can be run, the K8s cluster needs to understand what a test is and how to handle them. 
To do this TestSys needs to be installed to the cluster.
```bash
cli install --controller-uri controller:demo
```

### Run the test
The CLI installed earlier gives the ability to run tests on a cluster.
```bash

> cli run file example.yaml

Reviewer: When running this portion, I get the following error in the `hello-bones` pod and it fails:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Client(ClientError(Deserialization { source: Error("invalid type: map, expected u32", line: 0, column: 0) }))', agent/test-agent/examples/example_test_agent/main.rs:103:10
```

Is the `duplicator.info` selector valid? Maybe something good to add also would be some kind of "Troubleshooting" section of what to look at in the cluster to try and debug.

Author: The correct templating was `${duplicator.info}` instead of `{duplicator.info}`.

# Developer Guide for TestSys

## What is TestSys

> TestSys is a framework that leverages Kubernetes (k8s) to create resources and run tests.

Reviewer 2: super-duper nit: K8s
```suggestion
TestSys is a framework that leverages Kubernetes (K8s) to create resources and run tests.
```

To learn more about how it works, see the [design](design/DESIGN.md) document.

### Navigating the `bottlerocket-test-system` repo

> #### [Makefile.toml](Makefile.toml)

Reviewer 2: This should be `Makefile` since we don't use `cargo-make` in testsys (should we?)

> TestSys uses make to unify the logic needed to build images.

Reviewer 2: nit:
```suggestion
TestSys uses a makefile to unify the logic needed to build images.
```

I would even suggest removing this line, since the subsequent lines makes this point clear already.
```suggestion
```

> Any image can be built using `make <desired image>`.

Reviewer: ❓ Question: Is there a unified place that the different `<desired image>` are defined? I.e. is there a `make help` command or something we can point people to in order to discover what possible images can be created? (My guess is in the `bottlerocket-os/bottlerocket` docs?)

To make all bottlerocket related images, `make images` can be used, then `make tag-images` can be used to give all images the same tag and `make publish-images` can be used to publish all images.

#### [Dockerfile](Dockerfile)

> Testsys relies on docker images to run tests and create resources.

Reviewer 2: nit: I see "TestSys" used in some places, and "Testsys" elsewhere.

The Dockerfile contains the logic needed to create all bottlerocket related images.

#### [Agent](agent)

Testsys is broken up into 3 main things: the controller, test agent, and resource agent.

[agent-common](agent/agent-common) contains logic that is related to both the test and resource agents.

[test-agent](agent/test-agent) provides the framework for creating tests.
A test is a program that runs a test and reports its outcome by updating the TestSys Test CRD object's status fields. 
The test agent is packaged into a container image and the controller runs it in a pod as a Kubernetes Job. 

> To create a test agent, the [`Runner` trait](agent/test-agent/src/lib.rs#L51) needs to be implemented.

Reviewer 2: I would refrain from linking to a line number in a file since that can easily change. It would be more acceptable if the link is a permalink with specific commit ID info.
```suggestion
To create a test agent, the [`Runner` trait](agent/test-agent/src/lib.rs) needs to be implemented.
```

Check out the [example test agent](agent/test-agent/examples/example_test_agent/main.rs) to see how to create a test agent.

[resource-agent](agent/resource-agent) provides the framework for creating resources.

> A ResourceAgent can create and destroy external resources (such as Kubernetes clusters or compute instances) that are needed for a test. 

Reviewer 2: nit: markup for code-related terms and elsewhere as well
```suggestion
A `ResourceAgent` can create and destroy external resources (such as Kubernetes clusters or compute instances) that are needed for a test. 
```

We also refer to the test agent plainly as "test agent" right above, so this can just be 
```suggestion
A resource agent can create and destroy external resources (such as Kubernetes clusters or compute instances) that are needed for a test. 
```

The ResourceAgent program is packaged into a container image and is represented by a Resource CRD object.
A test can then depend on the existence of a resource and the controller will ensure the resource agent runs before the test is run. 
The resource agent is also responsible for destroying a resource after a test runs.
To create a resource agent, the [`Create` trait](agent/resource-agent/src/mod.rs#L37) and [`Destroy` trait](agent/resource-agent/src/mod.rs#L73) need to be implemented.
Check out the [example resource agent](agent/resource-agent/examples/example_resource_agent/main.rs) to see how to create a resource agent.

### [Bottlerocket](bottlerocket)

The bottlerocket directory of the repo is broken into 3 sections: agents, testsys, and types.

[agents](bottlerocket/agents) contains the implementations of all bottlerocket agents.

The types used by the agents can be found in [types](bottlerocket/types).

[testsys](bottlerocket/testsys) contains a cli implementation that is closely related to bottlerocket agents but will be removed in the near future.

> ### [Cli](cli)

Reviewer 2: `Cli` or should it be `CLI`?

For more info, see the [design](design/DESIGN.md) document. 

### [Model](model)

The model library contains the infrastructure that TestSys is built upon.
This includes the CRDs for [tests](model/src/test.rs) and [resources](model/src/resource.rs) as well as [everything](model/system) needed to be installed to the TestSys cluster.

All interactions with the TestSys cluster should be done using the model library through `TestManager`, `TestClient`, or `ResourceClient`.

The model provides [TestManager](model/src/test_manager/manager.rs) which can be used as an entry point for user interactions with TestSys.
To see how to use the TestManager look at the [example cli](cli).

Model also contains a [TestClient](model/src/clients/test_client) for simple interactions with Tests and a [ResourceClient](model/src/clients/resource_client) for Resources.

### [Selftest](selftest)

Selftest is a library used to validate incoming pull request by testing the overall TestSys framework.

## Getting Started

TestSys is currently tested on `x86_64` machines, these steps may not work as expected on other operating systems.

### Setting up TestSys

#### Create a TestSys directory

> Create a directory that you wish use as a working directory and navigate to it. 

Reviewer 3: s/wish use/wish to use/

The example provided creates a directory in the users home, but a directory may be created anywhere.

```bash
cd ~
mkdir testsys && cd testsys
pwd 
```

#### Clone the TestSys repo

> While in the TestSys directory, clone the `bottlerocket-test-system` repo from github.

Reviewer 3: I know this is a big nit, but should properly case company/product names like GitHub.

```bash
git clone https://github.com/bottlerocket-os/bottlerocket-test-system.git
cd bottlerocket-test-system
export TESTSYS_DIR=$(pwd)
```

The TestSys directory is now set up.

### Installing tools

#### Install `Kind`

> Instead of creating a long-standing eks cluster for TestSys development, `kind` enables the quick creation and destruction of local clusters that exist within the developers Docker which helps improve the testing experience.

Reviewer 2: nit:
```suggestion
Instead of creating a long-standing EKS cluster for TestSys development, `kind` enables the quick creation and destruction of local clusters that exist within the developers Docker which helps improve the testing experience.
```

and potentially elsewhere

Follow the [installation instructions](https://kind.sigs.k8s.io/docs/user/quick-start/) to set up `kind`.

#### Install `Kubectl`
While all testing user interactions can be done with `cli`, it is helpful to look at test, resource, and pod objects in a k8s cluster with `kubectl`.
Follow the [installation instructions]( https://kubernetes.io/docs/tasks/tools/) to set up `kubectl`.

#### Install Docker

> Docker is required for building TestSys images if docker is not installed follow the [installation instructions](https://docs.docker.com/get-docker/).

Reviewer 3: Just another nit:

s/images if docker/images. If docker/

```bash
echo '---
apiVersion: testsys.bottlerocket.aws/v1
kind: Test
metadata:
  name: hello-bones
  namespace: testsys-bottlerocket-aws
spec:
  agent:
    name: hello-agent
    image: "example-test-agent:demo"
    keepRunning: false
    configuration:
      mode: Fast
      person: Bones the Cat
      helloCount: 3
      helloDurationMilliseconds: 500
  resources: []
  dependsOn: []' > example_test_agent.yaml
```

### Create the `kind` cluster
Check if the `kind` cluster already exists.
```bash
kind get clusters
```

If the cluster already exists, it should be deleted.
```bash
kind delete cluster --name testsys-demo
```

Create the new demo cluster.
```bash
kind create cluster --name testsys-demo
```

Now the images created earlier need to be added to the cluster.
```bash
kind load docker-image controller:demo example-test-agent:demo --name testsys-demo
```

### Install TestSys
Before any tests can be run, the k8s cluster needs to understand what a test is and how to handle them. 
To do this TestSys needs to be installed to the cluster.
```bash
cli install --controller-uri controller:demo
```

### Run the test

> The cli installed earlier gives the ability to run tests on a cluster.

Reviewer 2: nit: "cli" in this context is an acronym so should be properly capitalized.
```suggestion
The CLI installed earlier gives the ability to run tests on a cluster.
```
and elsewhere as well.

```bash
cli logs --test hello-bones --follow
```

### Cleanup
When creating tests and resources it is important to maintain a clean state because tests must have a unique name in a cluster.
For instance, calling `cli run file example_test_agent.yaml` will cause an error because there is already a test named `hello-bones`.
To delete all tests and resources from the cluster use `cli delete`.
Another option is to remove everything from the TestSys namespace using `cli uninstall`.
This will delete the TestSys namespace and everything inside it, including the controller, secrets, tests and resources.

*Note: If a `kind` cluster is being created and destroyed each time, there is no need to manually delete the tests.*

### Delete the `kind` cluster
```bash
kind delete cluster --name testsys-demo
```

## Running a resource

### Navigate to the TestSys directory
```bash
cd ${TESTSYS_DIR}
```

### Build testing images
First, the images need to be created and tagged so they can be loaded into a `kind` cluster.
```bash
make controller
make example-test-agent
make duplicator-resource-agent
docker tag controller controller:demo
docker tag example-test-agent example-test-agent:demo

> docker tag duplicator-agent duplicator-agent:demo

Reviewer: This line fails since it appears `make duplicator-resource-agent` generates an image named `duplicator-resource-agent`

```
❯ docker images
REPOSITORY                 TAG               IMAGE ID       CREATED          SIZE
duplicator-resource-agent  latest            d513cbac4aca   11 minutes ago   33.8MB
```
